In [9]:
# We need this version or greater to have available the option "alternative" in the function ttest_1sample() from scipy
!pip install scipy==1.7.3 

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [10]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

In [11]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Please upload the file titanic_train.csv to your environment
data = pd.read_csv('/content/drive/MyDrive/CURR-v3.X-MAY2022/UNIT7/DAY2/7.04 inferential statistics/titanic_train.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [12]:
#Sample 30 passengers from 3rd class and get only the "Fare"
c3_sample = data[data['Pclass']==3]['Fare'].sample(30, random_state=1)
c3_sample.head()

642    27.900
171    29.125
280     7.750
354     7.225
590     7.125
Name: Fare, dtype: float64

$$H0: \mu ≥ 13.68$$
$$H1: \mu < 13.68$$

In [13]:
#compute the statistic and p-value

sample_mean = c3_sample.mean()
sample_std = c3_sample.std(ddof=1)
pop_mean = data[data['Pclass'] == 3]['Fare'].mean()
pop_std = data[data['Pclass'] == 3]['Fare'].std()
sem = sample_std/np.sqrt(30)
print("Population mean is: {:.2f}".format(pop_mean))
print("Population standard deviation is: {:.2f}".format(pop_std))
print()
print("Our sample mean is: {:.2f}".format(sample_mean))
print("Our sample standard deviation is: {:.2f}".format(sample_std))
print("The sem is: {:.2f}".format(sem))
print()
t = (sample_mean - pop_mean) / sem
print("Our statistic is: {:.2f}".format(t))
print("The p_value corresponding to our statistic is: {:.2f}".format(st.t.cdf(t,df = len(c3_sample)-1)))
print("The significance level is set to 0.05")
print("We accept the H0?")

Population mean is: 13.68
Population standard deviation is: 11.78

Our sample mean is: 12.90
Our sample standard deviation is: 10.22
The sem is: 1.87

Our statistic is: -0.42
The p_value corresponding to our statistic is: 0.34
The significance level is set to 0.05
We accept the H0?


Be careful here, because now our statistic t is:

$$t = \frac{(\bar{x} - \mu)}{\frac{\sigma}{\sqrt{n}}} < 0$$

-0.42 in our case.

Let's understand what happens. Acccording to null Hypothesis H0, the population mean should be bigger than $\mu = 13.68$. Therefore, any sample mean value bigger than $\mu$ will produce a positive value for the statistic (consistent with the H0). Thus, positive values in the distribution are allowed. Negative values will be associated with a sample mean lower than the population mean, which means H1. According to the t-Student distribution, all the values are possible although some of them are more likely than others. 

Here, we are wondering how likely is to get values smaller than 13.68. Beyond a negative value, the probability to get even smaller values with be really small.

Accodring to the t-Student distribution, the probability of getting any value smaller or equal to our statistic is given by the cumulative distribution function:

$$st.t.cdf(t) = 0.34$$

On the other hand, our significance value is 0.05. Therefore, as:

$$p-value = P(x <= t=-0.42) = 0.34 > 0.05=\alpha$$ 

we accept the H0

In [14]:
#0.34 > 0.05 
st.t.ppf(0.05,df= len(c3_sample)-1)

-1.6991270265334977

In [15]:
# start rejecting when (sample_mean - 13.68)/(sem) < -1.699

Another way to do the same t-test in a single shot.

In [16]:
# The meaning of the "alternative" keyword here is what is the alternative hypothesis H1.
# In our case, the alternative hypothesis is "<" -> less.
# Other possible values for the "alternative" keyword are:
# "two-sided"
# "greater"

t_statistic, p_value = st.ttest_1samp(c3_sample, popmean = 13.68, alternative = "less")
print("The t statistic of our sample is: {:.2f} and the corresponding p-value is: {:.2f}".format(t_statistic, p_value))

The t statistic of our sample is: -0.42 and the corresponding p-value is: 0.34


# Activity 1

Check if 1st class prices are different from 85 usd by sampling 30 1st class passengers and requiring a 5% significance.

Now we have:

$$H0: \mu = 85$$
$$H1: \mu \neq 85$$

Therefore, we have a two sided test.

In [17]:
c1_sample = data[data['Pclass'] == 1]['Fare'].sample(30,random_state=1)

In [18]:
pop_mean = 85
sample_mean = c1_sample.mean()
sample_std = c1_sample.std(ddof=1)
sem = sample_std/np.sqrt(30)
print("Population mean is: {:.2f}".format(85))
print()
print("Our sample mean is: {:.2f}".format(sample_mean))
print("Our sample standard deviation is: {:.2f}".format(sample_std))
print("The sem is: {:.2f}".format(sem))
print()
t = (sample_mean - pop_mean) / sem
print("Our statistic is: {:.2f}".format(t))
print("The p_value corresponding to our statistic is: {:.2f}".format(st.t.cdf(t,df = len(c1_sample)-1)))
print("The significance level is set to 0.05")
print("We accept the H0?")


Population mean is: 85.00

Our sample mean is: 81.19
Our sample standard deviation is: 104.59
The sem is: 19.10

Our statistic is: -0.20
The p_value corresponding to our statistic is: 0.42
The significance level is set to 0.05
We accept the H0?


Now we obtained that:

$$P(x<=t=-0.20)=0.42 > 0.05/2=0.025$$
$$P(x>t=-0.20) = 1-P(x<=t=0.20)=0.58 > 0.05/2=0.025$$

We accept the H0

In [19]:
# The meaning of the "alternative" keyword here is what is the alternative hypothesis H1.
# In our case, the alternative hypothesis is "<" -> less.
# Other possible values for the "alternative" keyword are:
# "two-sided"
# "greater"

t_statistic, p_value = st.ttest_1samp(c1_sample, popmean = 85, alternative = "two-sided")
print("The t statistic of our sample is: {:.2f} and the corresponding p-value is: {:.2f}".format(t_statistic, p_value))

The t statistic of our sample is: -0.20 and the corresponding p-value is: 0.84


In [ ]:
# End of activity

#Matched pairs

In [ ]:
#Example: Matched Pair.
#The data in the two samples is dependent. In the data we are studying the blood pressure before and after the treatment.

In [21]:
blood_pressure = pd.read_csv('/content/drive/MyDrive/CURR-v3.X-MAY2022/UNIT7/DAY2/7.04 inferential statistics/blood_pressure.csv')
blood_pressure.head()

before       after
0  136.713072   92.432965
1  134.735618  105.022643
2  127.529115   82.242766
3  144.527126   93.607172
4  124.214720  103.212223

Know we want to use an statistical test to know if the blood preasures after the treatment are significantly smaller that before.

For that we will compare the mean blood pressure before and after the treatment and we will compare them. Now we have

$$H0: \mu_{b} = \mu_{a} → \mu_{b} - \mu_{a} = 0$$
$$H1: \mu_{b} \neq \mu_{a} → \mu_{b} - \mu_{a} \neq 0$$

Given the unequality on the alternative, we need to use a "two-sided" test.

In [22]:
sample = blood_pressure.sample(30, random_state = 1)
sample['difference'] = sample['before']-sample['after']
sample.head()

before       after  difference
80  126.542579  115.228663   11.313916
84  133.586936  104.645561   28.941374
33  154.950211   93.409950   61.540262
81  125.917682   89.025141   36.892541
93  134.635064   99.986746   34.648319

In [23]:
sample_diff_mean, sample_diff_std = sample['difference'].mean(), sample['difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(38.384838963991704, 15.003794932150752)

Out t statistic is:

$$t = \frac{\overline{d}}{\frac{\sigma}{\sqrt{n}}}$$

where $\overline{d} = mean(sample_b - sample_a)$

In [24]:
t = sample_diff_mean / ( sample_diff_std / np.sqrt(sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 38.38
The standard deviation of our samples differences is: 15.00
Our t statistics is: 14.01


In [25]:
tc = st.t.ppf(1-(0.05/2),df= sample.shape[0] - 1)
tc

2.045229642132703

Our statistic is 14.01 while the critical value is 2.05. Then, as 14.01 > 2.05 we reject the H0. 

With p_values.

In [29]:
1-st.t.cdf(t,df = sample.shape[0] - 1)

9.658940314238862e-15

The probability to see a t value as big as this one is for the t-Student distribution is:

$$1-P(x <=t=14.01) < 0.025$$

we reject the H0.

In [27]:
# Another way.
st.ttest_rel(sample['before'],sample['after'])

Ttest_relResult(statistic=14.012616315976079, pvalue=1.9191321318300955e-14)

In [28]:
1-st.t.cdf(t,df = sample.shape[0] - 1)

9.658940314238862e-15

In [ ]:
#positive  statistic means \mu_before-\mu_after > 0   ---->  \mu_after < \mu_before 

# Pooled samples / independent samples

Now let's see what happens if we compare two different populations. The key point is that now we don't have a one to one correspondance between the samples before and after.

In [ ]:
#Independent Samples
#For two groups where we cannot match the observations to one another. In this case transactions from a website with different interfaces (a, b)

In [30]:
ab_test = pd.read_csv('/content/drive/MyDrive/CURR-v3.X-MAY2022/UNIT7/DAY2/7.04 inferential statistics/ab_test.csv')
ab_test.head()

a      b
0   0.27  13.61
1   6.08  21.53
2  13.74   9.23
3   9.70   5.36
4   7.00  12.90

Let's state our Hypothesis. Again we have

$$H0: \mu_{b} = \mu_{a} → \mu_{b} - \mu_{a} = 0$$
$$H1: \mu_{b} \neq \mu_{a} → \mu_{b} - \mu_{a} \neq 0$$

Given the unequality on the alternative, we need to use a "two-sided" test.

In [31]:
sample = ab_test.sample(30 , random_state = 1)
sample_a, sample_b = sample['a'], sample['b']

In this case, **if we assume that the variances are equal** (1/2 < s1/s2 < 2) the t statistic is given by:

$$t=\frac{(\mu_{a}-\mu_{b})}{s_{p}\sqrt{\frac{1}{n_{a}}+\frac{1}{n_{b}}}}$$

with:

$$s_{p}=\sqrt{\frac{(n_{a}-1)s_{a}^{2}+(n_{b}-1)s_{b}^{2}}{(n_{a}+n_{b}-2)}}$$

Let's compute the elements needed to obtain the statistic

In [33]:
sample_a_mean, sample_b_mean = sample['a'].mean(), sample['b'].mean()
sample_a_std, sample_b_std = sample['a'].std(ddof=1), sample['b'].std(ddof=1)
sp = ( len(sample_a) - 1 ) * ( sample_a_std**2 ) +  ( len(sample_b) - 1 ) * ( sample_b_std**2 )
sp /= ( len(sample_a) + len(sample_b) - 2)
sp = np.sqrt(sp)
r = np.sqrt( (1/len(sample_a)) + (1/len(sample_b)) )
t = ( sample_a_mean - sample_b_mean )/ (sp * r)

print("The mean of sample a is {:.2f}".format(sample_a_mean))
print("The mean of sample b is {:.2f}".format(sample_b_mean))
print("The standard deviation of sample a is {:.2f}".format(sample_a_std))
print("The standard deviation of sample b is {:.2f}".format(sample_b_std))
print("The ratio of the sample variances is {:.2f} which is bigger than 0.5 and smaller than 2".format(sample_a_std/sample_b_std))
print("The t statistic is: {:.2f}".format(t))

The mean of sample a is 9.04
The mean of sample b is 11.39
The standard deviation of sample a is 4.84
The standard deviation of sample b is 7.01
The ratio of the sample variances is 0.69 which is bigger than 0.5 and smaller than 2
The t statistic is: -1.51


Now let's do the test. We need to remember that this is a two-tailed test because the alternative hypothesis is an unequality.

Let's do the test first with critical values: 

In [34]:
st.t.ppf(1-(0.05/2),df = len(sample_a)+len(sample_b)-2)

2.0017174830120923

In [35]:
st.t.ppf((0.05/2),df = len(sample_a)+len(sample_b)-2)

-2.0017174830120927

The negative of the critical value is -2.00 and our t statistic is -0.21. Therefore, we accept the H0

Let's do the same with p_values:

In [ ]:
st.t.cdf(t,df = len(sample_a)+len(sample_b)-2)

0.0676447166798922

According to this, the probability of getting any value smaller or equal to our t-statistic is:

$$P(x<t=-1.51)=0.067 > 0.05/2=0.025$$

According to this, we will accept H0.

#ANOVA

This test takes samples of more than two different populations and test if all the sample means are the same or not (no matter which pair of means is different. Therefore, now our hypothesis is:

$$H0: \mu_{1}=\mu_{2}=\mu_{3}=⋯=\mu_{n}$$
$$H1: \mu_{i}\neq\mu_{j}$$

for any i,j


Keep in mind that we could also do C(N,2) two sample t-test for every pair of means.

In [ ]:
#ANOVA test
#to compare the means of multiple groups at the same time
#why not just multiple t-tests?

In [36]:
interest_r = pd.read_csv('/content/drive/MyDrive/CURR-v3.X-MAY2022/UNIT7/DAY2/7.04 inferential statistics/rate_by_city.csv')
interest_r.head()

Rate  City
0  13.75     1
1  13.75     1
2  13.50     1
3  13.50     1
4  13.00     1

In [37]:
interest_r['City'].value_counts()

1    9
2    9
3    9
4    9
5    9
6    9
Name: City, dtype: int64

In [38]:
group_df = interest_r.groupby('City')['Rate'].agg(City_mean='mean',Samples='size').reset_index()
group_df

City  City_mean  Samples
0     1  13.194444        9
1     2  12.611111        9
2     3  13.306667        9
3     4  13.244444        9
4     5  13.483333        9
5     6  12.200000        9

In case of the One way ANOVA test, the statistic to compute F is given by:

$$F = \frac{\hat{S}_{t}^{2}}{\hat{S}_{E}^{2}}$$

where:

$$\hat{S}_{t}^{2} = \frac{SST}{(K-1)}= \frac{\sum_{g}n_{g}(\bar{y}_{g}-\bar{y})^{2}}{(K-1)}$$

and:

$$\hat{S}_{E}^{2}=\frac{SSE}{(N-K)}=\frac{\sum_{g}\sum_{i}(y_{gi}-\bar{y}_{g})^{2}}{(N-K)}=\frac{\sum_{g}(n_{g}-1)s_{g}^{2}}{(N-K)}$$

In this expression $g$ means **group**, $i$ **value**, and $n_{g}$ is the number of observations in group $g$ and $K$ is the number of groups. Finally, the bar over $y$ means **mean**. 

These formulas looks horrible, but let's try to understand the terms. 

SST:

This term computes how much it **deviates each group mean from the global mean** and add the squares of those deviations multiplied by the number of members in the group divided by the number of groups K minus one.

SSE:

This other term, computes **how much every single value of every group deviates from the group mean**.

In summary, SST computes the variance of the group means from the global mean, while SSE computes the variance of the values against the global mean.

Let's start computing the needed terms to compute $\hat{S}_{t}^{2}$:

In [39]:
# In our case, the groups are the cities, and the "y" is the "Rate"
S2t = 0
for city in interest_r['City'].unique():
    ng = len(interest_r[interest_r['City'] == city])  
    S2t  += ng * ( ( interest_r[interest_r['City'] == city]['Rate'].mean() - interest_r['Rate'].mean() ) ** 2) 
S2t /= ( interest_r['City'].nunique() - 1 ) # Number of  K = groups/cities 

print("The value of S2t is {:.2f}".format(S2t)) 

The value of S2t is 2.19


Now let's go now with $\hat{S}_{E}^{2}$:

In [40]:
S2E = 0
for city in interest_r['City'].unique():
    for rate in interest_r[interest_r['City'] == city]['Rate']:
        S2E += ( rate - interest_r[interest_r['City'] == city]['Rate'].mean() ) ** 2
S2E /= ( len(interest_r) - interest_r['City'].nunique() ) 

print()
print("The value of S2E is {:.2f}".format(S2E))


The value of S2E is 0.45


In [41]:
S2E2 = 0
for city in interest_r['City'].unique():
    ng = len(interest_r[interest_r['City'] == city])
    S2E2 += ( ng - 1 ) * np.var(interest_r[interest_r['City'] == city]['Rate'],ddof=1)
S2E2 /= ( len(interest_r) - interest_r['City'].nunique() ) 

print()
print("The value of S2E is {:.2f}".format(S2E2))


The value of S2E is 0.45


Finally, we can compute F:

In [42]:
F = S2t / S2E
print("The value of F is {:.2f}".format(F))

The value of F is 4.83


The shape of the F distribution depends on two sets of degrees of freedom:  $d_{1}=K-1$ and $d_{2}=N-K$ 

In [43]:
interest_r['City'].nunique()

6

In [44]:
d1 = interest_r['City'].nunique() - 1
d2 = len(interest_r) - interest_r['City'].nunique()

print("Number of degrees of freedom d1: ",d1)
print("Number of degrees of freedom d2: ",d2)

Number of degrees of freedom d1:  5
Number of degrees of freedom d2:  48


The probability to get any F value lower or equal to out F can be obtained with the CDF:

In [45]:
st.f.cdf(F,dfn=d1, dfd=d2)

0.9988254485854959

Thus, the probability to get any value smaller or equal to F

$$P(x \le F=4.83)= 0.9988$$

The opposite is given by

In [46]:
1 - st.f.cdf(F,dfn=d1, dfd=d2)

0.0011745514145040659

Therefore, the probability to get a value bigger than F is:

$$P(x > F) = 1 - P(x \le F) = 0.001 < 0.05$$

Therefore, we reject the H0

The critical value which corresponds to an area of 0.05 is given by:

In [47]:
Fc = st.f.ppf(1-0.05,dfn=d1, dfd=d2)

print("The critical value which corresponds to an area of 0.05 is: {:.2f}".format(Fc))

The critical value which corresponds to an area of 0.05 is: 2.41


As our F is bigger than the critical value, we reject H0.

Also, scipy.stats includes a function to do the one way ANOVA at once.

In [48]:
print(st.f_oneway(interest_r[interest_r['City'] == 1]['Rate'],interest_r[interest_r['City'] == 2]['Rate'],
                  interest_r[interest_r['City'] == 3]['Rate'],interest_r[interest_r['City'] == 4]['Rate'],
                  interest_r[interest_r['City'] == 5]['Rate'],interest_r[interest_r['City'] == 6]['Rate']))

F_onewayResult(statistic=4.8293848737024, pvalue=0.001174551414504048)


Which gives us exactly what we had before!

As the p_value is < 0.05 we reject the H0.